sycl/source/detail/scheduler/commands.cpp

```diff
@@ -3011,6 +3023,89 @@ ur_result_t ExecCGCommand::enqueueImpCommandBuffer() {
     MEvent->setSyncPoint(OutSyncPoint);
     return UR_RESULT_SUCCESS;
   }
+  case CGType::EnqueueNativeCommand: {
+    // Queue is created by graph_impl before creating command to submit to
+    // scheduler.
+    const AdapterPtr &Adapter = MQueue->getAdapter();
+    auto ContextImpl = MQueue->getContextImplPtr();
+    auto DeviceImpl = MQueue->getDeviceImplPtr();
+    const auto Backend = DeviceImpl->getBackend();
+
+    // The CUDA & HIP backends don't have the equivalent of barrier
+    // commands that can be appended to the native UR command-buffer
+    // to enforce command-group predecessor and successor dependencies.
+    // Instead we create a new command-buffer to give to the user
+    // in the ext_codeplay_get_native_graph() call, which is then
+    // added to the main command-buffer in the UR cuda/hip adapter
+    // using cuGraphAddChildGraphNode/hipGraphAddChildGraphNode.
+    //
+    // As both the SYCL-RT and UR adapter need to have a handle
+    // to this child command-buffer, the SYCL-RT creates it and
+    // then passes the handle via a parameter to
+    // urCommandBufferAppendNativeCommandExp.
+    ur_exp_command_buffer_handle_t ChildCommandBuffer = nullptr;
+    if (Backend == sycl::backend::ext_oneapi_cuda ||
+        Backend == sycl::backend::ext_oneapi_hip) {
```

Comment:
Generally speaking, we try to do as little backend-specific logic in the runtime. Is there anything we can do to determine this from UR calls or is there a UR query we can add to make this possible?

Reply:
Good point, I've added a UR query for this `UR_DEVICE_INFO_COMMAND_BUFFER_SUBGRAPH_SUPPORT_EXP`